File: 15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from typing import List
# @lc code=start


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        size = len(nums)
        res = []
        if size < 3:
            return []

        def check(frist, second, third):
            if nums[first] + nums[second] + nums[third] == 0:
                res.append([nums[first], nums[second], nums[third]])

        for first in range(size):
            if first > 0 and nums[first] == nums[first - 1]:
                continue
            third = size - 1
            for second in range(first + 1, size):
                if second > first + 1 and nums[second] == nums[second - 1]:
                    continue
                while second < third and nums[first] + nums[second] + nums[third] > 0:
                    third -= 1
                if second == third:
                    break
                check(first, second, third)
        return res


# @lc code=end
if __name__ == "__main__":
    # a, b, c = map(int, input().split())
    # # 6种输出 但其实两种情况 分为a>b和a<=b
    # if a > b:
    #     if b > c:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(a, b, c))
    #     elif a > c:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(a, c, b))
    #     else:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(c, a, b))
    # else:
    #     if c > b:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(c, b, a))
    #     elif c > a:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(b, c, a))
    #     else:
    #         print("{:0.2f}>={:0.2f}>={:0.2f}".format(b, a, c))

    nums = list(map(int, input().split()))
    nums.sort(reverse=True)
    print("{:0.2f}>={:0.2f}>={:0.2f}".format(nums[0], nums[1], nums[2]))
    

File: 15.三数之和.java
import java.util.ArrayList;
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        if (n < 1)
            return res;
        Arrays.sort(nums);
        for (int i = 0; i < n; i++) {
            // 因为不允许重复的三元组
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int c = -nums[i];
            for (int j = i + 1, k = n - 1; j < k; j++) {
                // 对当前j找到一个最小的k使三个数加起来大于等于0
                if (j > i + 1 && nums[j] == nums[j - 1])
                    continue;
                while (j < k - 1 && nums[j] + nums[k-1] >= c) {
                    k--;
                }
                if (nums[j] + nums[k] == c) {
                    ArrayList lst = new ArrayList<>();
                    lst.add(nums[i]);
                    lst.add(nums[j]);
                    lst.add(nums[k]);
                    res.add(lst);
                }
            }
        }
        return res;
    }
}
// @lc code=end
